runtime/compiler/x/codegen/X86PrivateLinkage.cpp

```diff
@@ -2628,6 +2628,107 @@ void J9::X86::PrivateLinkage::buildVPIC(TR::X86CallSite &site, TR::LabelSymbol *
+ *    The register that will be used to load the interface class.
+ *
+ * \parm declaringClass
+ *    The class that is to be compared.
+ *
+ * \parm use32BitInterfacePointers
+ *    Whether or not using 32 bit interface class pointer.
+ *
+ * \parm lookupDispatchSnippetLabel
+ *    The label that dispatches to interface look up snippet if the comparison fails.
+ *
+ * \parm gotoLastITableDispatchLabel
+ *    The label that dispatches to invoke the interface call if the comparison succeeds.
+ *
+ * \parm comp
+ *    The compilation object
+ *
+ * \parm cg
+ *    The CodeGenerator object
+ *
+ */
+static void generateITableEntryCompare(uint32_t iterations,
+                                       TR::Node *callNode,
+                                       TR::Register *scratchReg,
+                                       TR::Register *vftReg,
+                                       TR::Register *interfaceClassReg,
+                                       TR_OpaqueClassBlock *declaringClass,
+                                       bool use32BitInterfacePointers,
+                                       TR::LabelSymbol *lookupDispatchSnippetLabel,
+                                       TR::LabelSymbol *gotoLastITableDispatchLabel,
+                                       TR_J9VMBase *fej9,
+                                       TR::Compilation *comp,
+                                       TR::CodeGenerator *cg)
+   {
+   for (uint32_t i = 0; i<iterations; i++)
+      {
+      // First entry
+      if (0 == i)
+         {
+         // scratchReg = j9class->iTable
+         generateRegMemInstruction(TR::InstOpCode::LRegMem(), callNode, scratchReg, generateX86MemoryReference(vftReg, offsetof(J9Class, iTable), cg), cg);
+         }
+      else
+         {
+         // scratchReg = iTable->next
+         generateRegMemInstruction(TR::InstOpCode::LRegMem(), callNode, scratchReg, generateX86MemoryReference(scratchReg, offsetof(J9ITable, next), cg), cg);
+         }
+
+      // Test if iTable == NULL?
+      generateRegRegInstruction(TR::InstOpCode::TESTRegReg(), callNode, scratchReg, scratchReg, cg);
+      generateLabelInstruction(TR::InstOpCode::JE4, callNode, lookupDispatchSnippetLabel, cg);
+
+      if (use32BitInterfacePointers)
+         {
+         // The field is 8 bytes, but only 4 matter
+         generateMemImmInstruction(TR::InstOpCode::CMP4MemImm4,
+                                  callNode,
+                                  generateX86MemoryReference(scratchReg, fej9->getOffsetOfInterfaceClassFromITableField(), cg),
+                                  (int32_t)(intptr_t)declaringClass,
+                                  cg);
+         }
+      else
+         {
+         TR_ASSERT_FATAL(comp->target().is64Bit(), "Only 64-bit path should reach here.");
+         auto cds = cg->findOrCreate8ByteConstant(callNode, (intptr_t)declaringClass);
+         TR::MemoryReference *interfaceClassAddr = generateX86MemoryReference(cds, cg);
```

Comment:
On 64-bit, it is more efficient to use an immediate load for this address.  Something like this:
```
generateRegImm64Instruction(TR::InstOpCode::MOV8RegImm64, callNode, interfaceClassReg, declaringClass, cg, reloKind);
```
In this case, `reloKind` should be `TR_ClassPointer`.  Whenever you embed a class constant in the code cache you need to make sure there is a relocation created for it.